### PR TITLE
SVE basic convert

### PR DESCRIPTION
**Pull request by jfalcou**

### CODE REVIEW

File: include/eve/detail/function/simd/arm/sve/friends.hpp [hunk 1]
```diff
@@ -44,15 +44,49 @@ requires sve_abi<abi_t<T, N>> { return svcmpge(sve_true<T>(), v, w); }
...
+  using abi_t = typename logical<wide<T,N>>::abi_type;
+  using abi_u = typename logical<wide<U,N>>::abi_type;
 
-template<real_value T, typename N>
+  callable_object<tag::convert_> const cvt{};
+
+  if constexpr(is_aggregated_v<abi_t> || is_aggregated_v<abi_u>)
+  {
+    // Clearly sub-optimal
+    return to_logical(v.mask() & cvt(w, as<logical<T>>{}).mask());
```
Comment (DenisYaroshevskiy) [inline include/eve/detail/function/simd/arm/sve/friends.hpp:60]: I think this should just slice if `is_aggregated_v<abi_t>`
Well, maybe we want to think about which one is smaller, but we can talk about that separately. Just slice, recurse, combine.

File: include/eve/module/core/regular/impl/abs.hpp [hunk 2]
```diff
@@ -36,12 +36,18 @@ abs_(EVE_SUPPORTS(cpu_), T const& a) noexcept
...
 abs_(EVE_SUPPORTS(cpu_), C const& cond, U const& t) noexcept
 {
   return mask_op(cond, eve::abs, t);
 }
+template<conditional_expr C, scalar_value U, typename N>
+EVE_FORCEINLINE auto
+abs_(EVE_SUPPORTS(cpu_), C const& cond, wide<U,N> const& t) noexcept
+{
+  return mask_op(cond, eve::abs, t);
+}
```
Comment (DenisYaroshevskiy) [inline include/eve/module/core/regular/impl/abs.hpp:50]: Why do you need this one now?

File: include/eve/module/core/regular/impl/simd/arm/sve/max.hpp [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+#include <eve/detail/implementation.hpp>
+
+namespace eve::detail
+{
+template<scalar_value T, typename N>
+EVE_FORCEINLINE auto
+max_(EVE_SUPPORTS(sve_), wide<T, N> const& a, wide<T, N> const& b) noexcept -> wide<T, N>
+requires sve_abi<abi_t<T, N>>
+{
+  return svmax_z(sve_true<T>(),a,b);
```
Comment (DenisYaroshevskiy) [inline include/eve/module/core/regular/impl/simd/arm/sve/max.hpp:22]: I think we have masked min/max

File: include/eve/module/core/regular/impl/simd/arm/sve/convert.hpp [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+      return wide<U, N>(t_t{svcvt_s64_z(mask, ll)}, t_t{svcvt_s64_z(mask, hh)});
+    else if constexpr( std::same_as<U,std::uint64_t> )
+      return wide<U, N>(t_t{svcvt_u64_z(mask, ll)}, t_t{svcvt_u64_z(mask, hh)});
+  };
+
+  if constexpr( sizeof(U) == 8 )  return sve_split(v);
+  else if constexpr( match(c_o, int32 ) ) return svcvt_s32_z(mask, v);
+  else if constexpr( match(c_o, uint32) ) return svcvt_u32_z(mask, v);
+  else return convert_impl(EVE_RETARGET(cpu_), v, tgt);
+}
```
Comment (DenisYaroshevskiy) [inline include/eve/module/core/regular/impl/simd/arm/sve/convert.hpp:49]: Please look at pack instructions.

File: include/eve/module/core/regular/impl/abs.hpp (hunk 2 above)
Comment (jfalcou) [reply include/eve/module/core/regular/impl/abs.hpp:50]: I thought I was clever but I was not :D
Fixed

File: include/eve/detail/function/simd/common/bitmask.hpp [hunk 5]
```diff
@@ -74,7 +74,8 @@ namespace eve::detail
       else
       {
         std::size_t res{0};
-        detail::apply<Wide::size()>( [&](auto... I) { ((res|=(p.get(I) << I)),...); });
+        using t_t = make_integer_t<std::max(Wide::size()/8,std::ptrdiff_t{1}),unsigned>;
+        detail::apply<Wide::size()>( [&](auto... I) { ((res|=(t_t{p.get(I)} << I)),...); });
```
[the remark this answers is not in the document]
Comment (jfalcou) [reply include/eve/detail/function/simd/common/bitmask.hpp:78]: I thought I was clever but I was not :D
Fixed

File: include/eve/module/core/regular/impl/abs.hpp (hunk 2 above)
Comment (jfalcou) [reply include/eve/module/core/regular/impl/abs.hpp:50]: It was indeed required then the other simplification made it uselesss. 
Removed

File: include/eve/detail/function/simd/arm/sve/friends.hpp (hunk 1 above)
Comment (jfalcou) [reply include/eve/detail/function/simd/arm/sve/friends.hpp:60]: Done.

File: include/eve/module/core/regular/impl/simd/arm/sve/max.hpp (hunk 3 above)
Comment (jfalcou) [reply include/eve/module/core/regular/impl/simd/arm/sve/max.hpp:22]: Yes but the masked binary function on SVE works slightly differently than avx512. I need more time to find otu how to use them.

File: include/eve/module/core/regular/impl/simd/arm/sve/convert.hpp (hunk 4 above)
Comment (jfalcou) [reply include/eve/module/core/regular/impl/simd/arm/sve/convert.hpp:49]: They only work for integers and do type promotion at the same time.
Next PR will cover them.

File: include/eve/detail/function/simd/arm/sve/friends.hpp [hunk 6]
```diff
@@ -44,15 +44,51 @@ requires sve_abi<abi_t<T, N>> { return svcmpge(sve_true<T>(), v, w); }
...
+  using abi_u = typename logical<wide<U,N>>::abi_type;
 
-template<real_value T, typename N>
+  callable_object<tag::convert_> const cvt{};
+
+  if constexpr(is_aggregated_v<abi_t> || is_aggregated_v<abi_u>)
+  {
+    auto[lv,hv] = v.slice();
+    auto[lw,hw] = w.slice();
+    return logical<wide<T, N>>{ lv || lw, hv || hw};
```
Comment (DenisYaroshevskiy) [inline include/eve/detail/function/simd/arm/sve/friends.hpp:61]: If left is not aggregated the else block is a better option I think

File: include/eve/detail/function/simd/arm/sve/friends.hpp [hunk 7]
```diff
@@ -44,15 +44,45 @@ requires sve_abi<abi_t<T, N>> { return svcmpge(sve_true<T>(), v, w); }
...
 EVE_FORCEINLINE auto
-self_logand(sve_ const&, logical<wide<T, N>> v, logical<wide<T, N>> w) noexcept -> logical<wide<T, N>>
-requires sve_abi<abi_t<T, N>> { return svmov_z(v, w); }
+self_logand(sve_ const&, logical<wide<T,N>> v, logical<wide<U,N>> w) noexcept -> logical<wide<T, N>>
+requires(sve_abi<abi_t<T, N>> || sve_abi<abi_t<U, N>>)
+{
+  using abi_t = typename logical<wide<T,N>>::abi_type;
 
-template<real_value T, typename N>
+  if constexpr(!is_aggregated_v<abi_t>)
```
Comment (DenisYaroshevskiy) [inline include/eve/detail/function/simd/arm/sve/friends.hpp:54]: Just say abi_t<T, N>

**Review by DenisYaroshevskiy — APPROVED** [review]

**Pull request merged**